compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -62,21 +62,13 @@ struct ConstToPat<'tcx> {
     treat_byte_string_as_slice: bool,
 }
 
-mod fallback_to_const_ref {
-    #[derive(Debug)]
-    /// This error type signals that we encountered a non-struct-eq situation behind a reference.
-    /// We bubble this up in order to get back to the reference destructuring and make that emit
-    /// a const pattern instead of a deref pattern. This allows us to simply call `PartialEq::eq`
-    /// on such patterns (since that function takes a reference) and not have to jump through any
-    /// hoops to get a reference to the value.
-    pub(super) struct FallbackToConstRef(());
-
-    pub(super) fn fallback_to_const_ref(c2p: &super::ConstToPat<'_>) -> FallbackToConstRef {
-        assert!(c2p.behind_reference.get());
-        FallbackToConstRef(())
-    }
-}
-use fallback_to_const_ref::{fallback_to_const_ref, FallbackToConstRef};
+/// This error type signals that we encountered a non-struct-eq situation.
+/// We bubble this up in order to get back to the reference destructuring and make that emit
+/// a const pattern instead of a deref pattern. This allows us to simply call `PartialEq::eq`
+/// on such patterns (since that function takes a reference) and not have to jump through any
+/// hoops to get a reference to the value.
```

Comment:
I am confused by this comment. The error is consumed [here](https://github.com/rust-lang/rust/blob/9ff2e4e5d1802af102380b34162d120505a13862/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs#L139-L145), if I understand correctly, and I don't see anything there that would be specific to references. I don't see anything about this error code at all that refers to references. It gets raised when we encounter an ADT that is not structural_eq, and then means that the entire constant is treated opaquely, and that's that.

is the part about references (in the comments and the name) a leftover from earlier times?

Reply:
Ah, scratch that, I found the `self.behind_reference.get()`.

Otherwise we return `PatKind::Wild`? How does that make sense? If we compile this pattern, it will match everything...?? I guess it's okay because we did `emit_err`. But there's not even a comment there explaining that this pattern is completely irrelevant, so I am a bit confused.^^

Reply:
That confused me too, because this occasionally triggers spurious "unreachable pattern" warnings. I didn't manage to get rid of them at the time

Reply:
@Nadrieril shouldn't it be possible to get rid of them by returning a `Constant` pattern instead of a `Wild` pattern here?

Reply:
Yep that should do the trick, it would be treated as opaque by exhaustiveness checking

Reply:
https://github.com/rust-lang/rust/pull/115937 addresses this